### PR TITLE
Class details page

### PR DESCRIPTION
**Pull request by cindy-wang328**
https://social-diversity-for-children-pr-46.up.railway.app/en/programdetails/1

image

### CODE REVIEW

**Comment by cindy-wang328** [comment]
image

File: src/components/ClassList.tsx [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                                        <Box fontWeight="bold" as="h2">
+                                            {item.name}
+                                        </Box>
+                                        <Box
+                                            as="span"
+                                            color="gray.600"
+                                            fontSize="sm"
+                                            textTransform="capitalize"
+                                        >
+                                            {item.weekday
```
Comment (feng-eric) [inline src/components/ClassList.tsx]: Just a suggestion but maybe we can make this a function since we might use it elsewhere too? Also would improve code readability.

File: src/components/ClassList.tsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                                                .charAt(0)
+                                                .toUpperCase() +
+                                                item.weekday
+                                                    .substr(1)
+                                                    .toLowerCase()}{" "}
+                                            {(item.startTimeMinutes / 60) % 12}
```
Comment (feng-eric) [inline src/components/ClassList.tsx]: Can you leave some comments just so it's easier to understand all the expressions

File: src/components/ProgramDetailsPage.tsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            <Text as="span" color="gray.600" fontSize="sm" mt="5">
+                {programInfo.startDate} to {programInfo.endDate}
+            </Text>
+            <Text mt="5">{programInfo.description}</Text>
+            <Flex mt="5" align="center">
+                <Text fontSize="sm" fontWeight="semibold">
+                    Select a class
+                </Text>
+                <Spacer />
+                {/* TODO what is the filter button supposed to do? */}
```
Comment (feng-eric) [inline src/components/ProgramDetailsPage.tsx]: Is this a question for designers?

Comment (feng-eric) [reply src/components/ProgramDetailsPage.tsx]: @stacykwok @christinaruu

**Review by feng-eric — COMMENTED** [review]
Left some comments!

File: src/components/ProgramDetailsPage.tsx (hunk 3 above)
Comment (cindy-wang328) [reply src/components/ProgramDetailsPage.tsx]: There was a filter button on the figma but i didn't see what it should do / what filter options there are (sorry if i missed it!)
image

File: src/components/ClassList.tsx (hunk 2 above)
Comment (cindy-wang328) [reply src/components/ClassList.tsx]: The substr and toLower/Upper Case is from https://stackoverflow.com/questions/196972/convert-string-to-title-case-with-javascript, to convert the weekday enums (MON, TUE, WED...) to Mon, Tue, Wed etc
The / 60 % 12 is to convert time in minutes to am/pm dividing by 60

File: src/components/ClassList.tsx [hunk 4]
```diff
@@ -17,6 +17,15 @@ import type { ClassCardInfo } from "models/Class";
 export const ClassList: React.FC<{ classInfo: ClassCardInfo[] }> = ({
     classInfo,
 }) => {
+    function convertToAmPm(timeInMinutes: Int) {
```
Comment (feng-eric) [inline src/components/ClassList.tsx]: did you mean to put `number` as the type?

Comment (feng-eric) [reply src/components/ClassList.tsx]: Let's also add some documentation comments on helper functions to maintain a standard!

Comment (cindy-wang328) [reply src/components/ClassList.tsx]: LMAO vscode so laggy it didnt catch that

File: src/components/ClassList.tsx [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    // time in minutes is minutes from 0:00, convert to something like "1:23 am"
+    function convertToAmPm(timeInMinutes: number) {
+        return (
+            (Math.floor(timeInMinutes / 60) % 12).toString() +
+            ":" +
+            ("0" + (timeInMinutes % 60)).slice(-2) +
+            (timeInMinutes >= 720 ? "pm" : "am")
+        );
+    }
+    //  from https://stackoverflow.com/questions/196972/convert-string-to-title-case-with-javascript
```
Comment (feng-eric) [inline src/components/ClassList.tsx]: @mahadzaryab1 Is there a standard you would prefer for code documentation for helper functions like these

**Comment by feng-eric** [comment]
@cindy-wang328 Out of the 2 PRs you opened, which one should we merge in first. I'm assuming after one is merged in, you'll remove the test data?

**Comment by cindy-wang328** [comment]
> @cindy-wang328 Out of the 2 PRs you opened, which one should we merge in first. I'm assuming after one is merged in, you'll remove the test data?

I thought removing test data counts as integration and was gonna be a separate task? The only thing that might be affected rn is, one test data in this pr can be specified to be ProgramCardInfo type after the other one is merged. But maybe that can just be part of the integration / removing test data

**Comment by cindy-wang328** [comment]
image
Added routing and changed the hardcoded test data (http://localhost:3000/en/programdetails/2 or /1) 
Fixed 12Am/12PM edge case in time function
Changed badge color

File: src/components/ClassList.tsx [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                                            pt="1.5"
+                                            px="3"
+                                        >
+                                            {item.ageGroup}
+                                        </Badge>
+                                        <Spacer />
+                                        <Box
+                                            as="span"
+                                            color="gray.600"
+                                            fontSize="sm"
```
Comment (jason17h) [inline src/components/ClassList.tsx]: For the participant availability text, could you set the `textAlign` prop to `right` and maybe reduce the vertical spacing a little bit? https://www.figma.com/file/ad7RNFErTGP7HkL52edTwQ/SDC-External-Application?node-id=1046%3A14307

File: src/components/ClassList.tsx [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                                            alt={item.name}
+                                        />
+                                    </AspectRatio>
+                                </GridItem>
+                                <GridItem colSpan={3}>
+                                    <VStack
+                                        align="left"
+                                        justify="center"
+                                        height="100%"
+                                    >
```
Comment (jason17h) [inline src/components/ClassList.tsx:60]: Very small comment but could you change the info text to be in the format of 
`Mondays 6:00pm - 7:00pm`
`Teacher: Brian`
to be more similar to the figma?

**Review by jason17h — COMMENTED** [review]
Looks really good!!! I just had some minor comments about the styling and wrapping it with the navbar and footer but other than that it looks great

File: pages/programdetails/[pid].tsx [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+import React from "react";
+import { Heading, Flex, Badge, Spacer, Text, Button } from "@chakra-ui/react";
+import type { ClassCardInfo } from "models/Class";
+import { ClassList } from "src/components/ClassList";
+import { weekday } from "@prisma/client";
+
+export const ProgramDetails: React.FC = () => {
+    const router = useRouter();
+    const { pid } = router.query;
+    const programName: string = typeof pid === "string" ? pid : "";
```
[the remark this answers is not in the document]
Comment (cindy-wang328) [reply pages/programdetails/[pid].tsx:14]: It is "string | string[]" and if i didn't do that then it didn't let me use as a hashmap key
i tried searching stackoverflow and it looked like some built in next js feature

File: pages/programdetails/[pid].tsx [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+                volunteerSpaceAvailable: 4,
+                weekday: weekday.THU,
+                startTimeMinutes: 765,
+                durationMinutes: 60,
+                teacherName: "Teacher2",
+            },
+        ],
+    };
+    // TODO this will be a ProgramCardInfo (once the reverted PR gets unreverted)
+    const programInfo: { [programId: string]: any } = {
```
[the remark this answers is not in the document]
Comment (cindy-wang328) [reply pages/programdetails/[pid].tsx:67]: Right now it's actually the index of the card in the #48 
I think it should eventually be the PID? It doesn't matter too much if we're taking out the test data later? I didn't research how to get stuff from the backend yet tho, i guess for test data purposes the id is the index LOL

File: src/components/ClassList.tsx (hunk 6 above)
Comment (cindy-wang328) [reply src/components/ClassList.tsx]: 😭  i saw that but didnt know how to do it bc then i cant use grid 
edit: nvm i can use flex and spacer

File: pages/programdetails/[pid].tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (feng-eric) [reply pages/programdetails/[pid].tsx:67]: Ah ok yeah if we're taking it out later than this is fine to keep like this

File: src/components/ClassList.tsx (hunk 7 above)
Comment (cindy-wang328) [reply src/components/ClassList.tsx:60]: how do you want to map the weekday enum to "Mondays" then? I thought of putting a dict for it but then thought it would be bad reusability

**Comment by cindy-wang328** [comment]
image
added wrapper

File: src/components/ClassList.tsx (hunk 6 above)
Comment (jason17h) [reply src/components/ClassList.tsx]: omg sorry I don't know anything about flex container spacer etc. but I'm glad you figured it out (??)

**Comment by cindy-wang328** [comment]
image
added util function to convert weekday to string

**Review by feng-eric — APPROVED** [review]
LGTM!

**Review by jason17h — APPROVED** [review]
Looks really good! Only thing to pay attention to is if https://github.com/uwblueprint/social-diversity-for-children/pull/52 gets merged before this, you might need to remove the `48` padding since it got added to the `SDCWrapper` component. Otherwise, I think it's good to go!
